app/src/main/java/edu/usf/sas/pal/muser/util/HeadphoneUtils.kt

```diff
@@ -0,0 +1,111 @@
+             val volumeLevel = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC)
+             val volumeMax = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
+             var volumeMin = Int.MIN_VALUE
+             var volumeLevelDB = Float.MIN_VALUE
+             if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                 volumeMin = audioManager.getStreamMinVolume(AudioManager.STREAM_MUSIC)
+                 volumeLevelDB = if (isSpeakerOn(context)) {
+                     audioManager.getStreamVolumeDb(AudioManager.STREAM_MUSIC,
+                             volumeLevel, AudioDeviceInfo.TYPE_BUILTIN_SPEAKER)
+                 } else {
+                     audioManager.getStreamVolumeDb(AudioManager.STREAM_MUSIC,
+                             volumeLevel, deviceType)
+                 }
+             }
+             return VolumeData(volumeLevel, volumeMax, volumeMin, volumeLevelDB)
+         }
+        return null
+     }
+
+    /**
+     * Function to check if the built in speaker is used or not
+     * @param context - context of the activity/fragment
+     * @return isSpeakerOn true or false.
+     */
+    @JvmStatic
+    fun isSpeakerOn(context: Context): Boolean{
+        val headsetTypes: Array<Int>
+        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager?
+        var isSpeakerOn = true
+        if (audioManager != null) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                //  AudioDeviceInfo.TYPE_USB_HEADSET is only supported for devices with API_LEVEL > 26
+                headsetTypes = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                    arrayOf(AudioDeviceInfo.TYPE_BLUETOOTH_A2DP, AudioDeviceInfo.TYPE_BLUETOOTH_SCO,
+                            AudioDeviceInfo.TYPE_WIRED_HEADSET, AudioDeviceInfo.TYPE_WIRED_HEADPHONES,
+                            AudioDeviceInfo.TYPE_USB_HEADSET, AudioDeviceInfo.TYPE_USB_DEVICE)
+                } else {
+                    arrayOf(AudioDeviceInfo.TYPE_BLUETOOTH_A2DP, AudioDeviceInfo.TYPE_BLUETOOTH_SCO,
+                            AudioDeviceInfo.TYPE_WIRED_HEADSET, AudioDeviceInfo.TYPE_WIRED_HEADPHONES,
+                            AudioDeviceInfo.TYPE_USB_DEVICE)
+                }
+                val audioDevices = audioManager.getDevices(AudioManager.GET_DEVICES_OUTPUTS)
+                if (audioDevices != null) {
+                    for (audioDevice in audioDevices) {
+                        Log.d(TAG, "isSpeakerOn: ${audioDevice.type}")
+                        if (headsetTypes.contains(audioDevice.type)) {
+                            deviceType = audioDevice.type
+                            isSpeakerOn = false
+                            break
+                        }
+                    }
+                }
+            }
+            else {
+                isSpeakerOn = !audioManager.isWiredHeadsetOn && !audioManager.isWiredHeadsetOn
```

Comment:
Looks like `isWiredHeadsetOn` is deprecated - docs say:

>This is not a valid indication that audio playback is actually over the wired headset as audio routing depends on other conditions

Is the "right" way to do this the first part of the IF branch for SDK_INT <= M?

Also, you have ` !audioManager.isWiredHeadsetOn` twice here.

Reply:
After researching this on StackOverflow and other forums, looks like other developers preferred to do it this way. Android doesn't have an explicit function yet for API_LEVEL >= M, so I think this may be the best way to do this.